### PR TITLE
text api changes

### PR DESCRIPTION
**Pull request by szha**
## Description ##
host embeddings from s3. fixes.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Serve embedding files from apache-mxnet s3 bucket backed by CDN. Include licenses in doc.
- [x] Include all 294 fast-text embedding files as built-ins.
- [x] Refactor download logic. Reuse mxnet repo logic from other modules.
- [x] Check existing embedding files for hash match and skip download if matched
- [x] Update embedding parsing to be line-by-line to reduce peak memory usage.
- [x] Update test for downloadable embedding to not rely on 800mb file, and use relative path instead of default user path in tests.

## Comments ##
- No API change.

### CODE REVIEW

**Comment by szha** [comment]
@astonzhang

File: python/mxnet/text/embedding.py [hunk 1]
```diff
@@ -462,6 +465,10 @@ class GloVe(TokenEmbedding):
     To get the updated URLs to the externally hosted pre-trained token embedding
     files, visit https://nlp.stanford.edu/projects/glove/
 
+    License for pretrained embeddings:
```
Comment (astonzhang) [inline python/mxnet/text/embedding.py]: pre-trained

File: python/mxnet/text/embedding.py [hunk 2]
```diff
@@ -496,44 +503,27 @@ class GloVe(TokenEmbedding):
     """
 
     # Map a pre-trained token embedding archive file and its SHA-1 hash.
-    pretrained_archive_name_sha1 = \
-        {'glove.42B.300d.zip': 'f8e722b39578f776927465b71b231bae2ae8776a',
-         'glove.6B.zip': 'b64e54f1877d2f735bdd000c1d7d771e25c7dfdc',
-         'glove.840B.300d.zip': '8084fbacc2dee3b1fd1ca4cc534cbfff3519ed0d',
-         'glove.twitter.27B.zip': 'dce69c404025a8312c323197347695e81fd529fc'}
+    pretrained_archive_name_sha1 = C.GLOVE_PRETRAINED_FILE_SHA1
```
Comment (astonzhang) [inline python/mxnet/text/embedding.py]: We can replace pretrained_archive_name_sha1 with C.GLOVE_PRETRAINED_FILE_SHA1. There is no need for an extra var called pretrained_archive_name_sha1.

File: python/mxnet/text/embedding.py [hunk 3]
```diff
@@ -496,44 +503,27 @@ class GloVe(TokenEmbedding):
...
-             '767d80889d8c8a22ae7cd25e09d0650a6ff0a502',
-         'glove.twitter.27B.50d.txt':
-             '9585f4be97e286339bf0112d0d3aa7c15a3e864d',
-         'glove.twitter.27B.100d.txt':
-             '1bbeab8323c72332bd46ada0fc3c99f2faaa8ca8',
-         'glove.twitter.27B.200d.txt':
-             '7921c77a53aa5977b1d9ce3a7c4430cbd9d1207a'}
-
-    url_prefix = 'http://nlp.stanford.edu/data/'
+    pretrained_file_name_sha1 = C.GLOVE_PRETRAINED_ARCHIVE_SHA1
```
Comment (astonzhang) [inline python/mxnet/text/embedding.py]: Same as above. There is no need for an extra variable called pretrained_file_name_sha1

File: python/mxnet/text/embedding.py [hunk 4]
```diff
@@ -605,22 +599,16 @@ class FastText(TokenEmbedding):
     """
 
     # Map a pre-trained token embedding file and its SHA-1 hash.
-    pretrained_file_name_sha1 = \
-        {'wiki.en.vec': 'c1e418f144ceb332b4328d27addf508731fa87df',
-         'wiki.simple.vec': '55267c50fbdf4e4ae0fbbda5c73830a379d68795',
-         'wiki.zh.vec': '117ab34faa80e381641fbabf3a24bc8cfba44050'}
-    url_prefix = 'https://s3-us-west-1.amazonaws.com/fasttext-vectors/'
+    pretrained_file_name_sha1 = C.FAST_TEXT_FILE_SHA1
```
Comment (astonzhang) [inline python/mxnet/text/embedding.py]: Same as above. There is no need for the extra var

File: python/mxnet/text/embedding.py [hunk 5]
```diff
@@ -88,18 +88,24 @@ def __init__(self, **kwargs):
         super(TokenEmbedding, self).__init__(**kwargs)
 
     @classmethod
-    def _get_pretrained_file_path_from_url(cls, url, embedding_root,
-                                           pretrained_file_name):
-        """Get the local path to the pre-trained token embedding file from url.
+    def _get_download_file_name(cls, pretrained_file_name):
```
Comment (astonzhang) [inline python/mxnet/text/embedding.py]: It looks that argument pretrained_file_name is returned. Why do we need this method?

File: python/mxnet/text/embedding.py [hunk 6]
```diff
@@ -88,18 +88,24 @@ def __init__(self, **kwargs):
...
-        been downloaded yet or the existing file fails to match its expected
-        SHA-1 hash.
-        """
+        url_format = '{repo_url}gluon/embeddings/{cls}/{file_name}'
+        return url_format.format(repo_url=repo_url,
+                                 cls=embedding_cls,
+                                 file_name=cls._get_download_file_name(pretrained_file_name))
 
+    @classmethod
+    def _get_pretrained_file(cls, embedding_root, pretrained_file_name):
```
Comment (astonzhang) [inline python/mxnet/text/embedding.py]: I think this method name may be confusing: does the "pre-trained file" also cover a custom pre-trained file? How about def _get_pretrained_file_from_url

File: python/mxnet/text/embedding.py [hunk 7]
```diff
@@ -88,18 +88,24 @@ def __init__(self, **kwargs):
...
+        embedding_cls = cls.__name__.lower()
 
-        The pre-trained embedding file will be downloaded from url if it has not
-        been downloaded yet or the existing file fails to match its expected
-        SHA-1 hash.
-        """
+        url_format = '{repo_url}gluon/embeddings/{cls}/{file_name}'
+        return url_format.format(repo_url=repo_url,
+                                 cls=embedding_cls,
+                                 file_name=cls._get_download_file_name(pretrained_file_name))
```
Comment (astonzhang) [inline python/mxnet/text/embedding.py]: file_name is unused. Please remove it.

File: python/mxnet/text/constants.py [hunk 8]
```diff
@@ -22,3 +22,323 @@
...
+     'wiki.pnb.vec': '35f38862d3d83012d6db7baa8a4105e3e0a416e7',
+     'wiki.wo.vec': '2ad96a7a9e640bc0dbcf316b1f414b92802dcb8e',
+     'wiki.wuu.vec': 'e1cbae1d3ad52329d0f36ada764016fbacf07049',
+     'wiki.xh.vec': 'bf37f741b0b75953281d11df2b4d80100df9e666',
+     'wiki.yi.vec': '299d61958b7dcc38774768f1489121384726d860',
+     'wiki.yo.vec': 'e35c8aff2924ba07936be9d0d94bd298f09702a4',
+     'wiki.diq.vec': '77f3c370d1d77806fafe368cf788af550ff607dd',
+     'wiki.zea.vec': 'ee12db26aab3f2b3b2745a298ef414e7aeb5a058',
+     'wiki.za.vec': 'e3a0e58bd2e5b1891c71f1f7e37ff71997a20361',
+     'wiki.zu.vec': '4b244b9697a8280e6646842c5fc81bb3a6bc8ec7'}
```
Comment (astonzhang) [inline python/mxnet/text/constants.py]: Thank you for providing full set of fastText embeddings. Could you please test all the added pre-trained files? For context, wiki.simple, wiki.en, and wiki.zn have been tested.

Comment (astonzhang) [reply python/mxnet/text/constants.py]: Just to make sure that all these files are working properly.

File: tests/python/unittest/test_text.py [hunk 9]
```diff
@@ -119,49 +119,33 @@ def test_indices_to_tokens():
...
-    assert_almost_equal(unk_vec_sum, 0)
+            self._load_embedding(pretrained_file_path, ' ', init_unknown_vec)
 
-    unk_vecs_sum = glove_6b_50d.get_vecs_by_tokens(
-        ['<unk$unk@unk>', '<unk$unk@unk>']).sum().asnumpy()[0]
-    assert_almost_equal(unk_vecs_sum, 0)
+    test_embed = TokenEmbedding.create('test')
+    assert test_embed.token_to_idx['hello'] == 1
+    assert test_embed.token_to_idx['world'] == 2
+    assert_almost_equal(test_embed.idx_to_vec[1].asnumpy(), (nd.arange(5)+1).asnumpy())
```
Comment (astonzhang) [inline tests/python/unittest/test_text.py]: nd.arange(5) + 1 (space)

File: tests/python/unittest/test_text.py [hunk 10]
```diff
@@ -119,49 +119,33 @@ def test_indices_to_tokens():
...
+    assert_almost_equal(test_embed.idx_to_vec[2].asnumpy(), (nd.arange(5)+6).asnumpy())
```
Comment (astonzhang) [inline tests/python/unittest/test_text.py]: nd.arange(5) + 6 (space)

File: tests/python/unittest/test_text.py [hunk 11]
```diff
@@ -119,49 +119,33 @@ def test_indices_to_tokens():
...
+    @TokenEmbedding.register
+    class Test(TokenEmbedding):
+        pretrained_file_name_sha1 = \
+            {'embedding_test.vec': '29b9a6511cf4b5aae293c44a9ec1365b74f2a2f8'} # 33 bytes
+        namespace = 'test'
 
-def test_glove():
-    glove_6b_50d = TokenEmbedding.create(
-        'glove', pretrained_file_name='glove.6B.50d.txt')
+        def __init__(self, embedding_root=os.path.join('~', '.mxnet', 'embeddings'),
```
Comment (astonzhang) [inline tests/python/unittest/test_text.py]: embedding_root='embeddings'

File: tests/python/unittest/test_text.py [hunk 12]
```diff
@@ -119,49 +119,33 @@ def test_indices_to_tokens():
...
 
-    unk_vec_sum = glove_6b_50d.get_vecs_by_tokens(
-        '<unk$unk@unk>').sum().asnumpy()[0]
-    assert_almost_equal(unk_vec_sum, 0)
+            self._load_embedding(pretrained_file_path, ' ', init_unknown_vec)
 
-    unk_vecs_sum = glove_6b_50d.get_vecs_by_tokens(
-        ['<unk$unk@unk>', '<unk$unk@unk>']).sum().asnumpy()[0]
-    assert_almost_equal(unk_vecs_sum, 0)
+    test_embed = TokenEmbedding.create('test')
```
Comment (astonzhang) [inline tests/python/unittest/test_text.py]: To be consistent with common use cases, such as
glove_6b_50d = TokenEmbedding.create('glove', pretrained_file_name='glove.6B.50d.txt')
fasttext_simple = TokenEmbedding.create('fasttext', pretrained_file_name='wiki.simple.vec', init_unknown_vec=nd.ones)

Can we provide both arguments embedding_name ('test') + pretrained_file_name ('embedding_test.vec').

**Comment by astonzhang** [comment]
@szha In addition, please fix the failed tests. Thanks.

**Comment by astonzhang** [comment]
@szha  As a side note, although it saves time to test a dummy hosted "Test" embedding, our provided GloVe and FastText class have no test coverage. I feel it would be better to still provide such coverage, say the smallest pre-trained files (in total take ~60 secs at my laptop):

glove_6b_50d = TokenEmbedding.create('glove', pretrained_file_name='glove.6B.50d.txt', embedding_root='embeddings')

fasttext_simple = TokenEmbedding.create('fasttext', pretrained_file_name='wiki.simple.vec', init_unknown_vec=nd.ones, embedding_root='embeddings')

What do you think?

File: python/mxnet/text/embedding.py (hunk 5 above)
Comment (szha) [reply python/mxnet/text/embedding.py]: This is the mapping from the file-to-load to the file-to-download. Here in the base class, it assumes those two are the same. This function is overridden in GloVe because files are in zipped archives.

File: python/mxnet/text/embedding.py (hunk 7 above)
Comment (szha) [reply python/mxnet/text/embedding.py]: ?

File: python/mxnet/text/embedding.py (hunk 6 above)
Comment (szha) [reply python/mxnet/text/embedding.py]: That’s too long and binds url in the method. After the change, URL is no longer in argument.

File: python/mxnet/text/embedding.py (hunk 2 above)
Comment (szha) [reply python/mxnet/text/embedding.py]: Isn’t that the interface? Remember hasattr?

File: tests/python/unittest/test_text.py (hunk 12 above)
Comment (szha) [reply tests/python/unittest/test_text.py]: This test is intended for testing the download and integrity verification logic. Whether it has pretrained file name as argument or hard-coded shouldn't affect the result.

File: python/mxnet/text/constants.py (hunk 8 above)
Comment (szha) [reply python/mxnet/text/constants.py]: I've already tested them after uploading. For obvious reason we can't include the complete set as part of the CI tests.
